File: bench/manual/Hanoi.java

public class Hanoi {
    private int moves;
    public Hanoi() {
        this.moves = 0;
    }

    private void doHanoi(int n, int to, int from, int using) {
        if(n > 0) {
            doHanoi(n - 1, using, from, to);
            moves++;
            doHanoi(n - 1, to, using, from);
        }
    }

    public void hanoi(int n) {
        doHanoi(n, 3, 1, 2);
        System.out.println("" + moves + " moves total");
    }

    public static void main(String[] args) throws Exception {
        new Hanoi().hanoi(20);
    }
}

File: bench/manual/hanoi.rb

class Hanoi
  def initialize
    @moves = 0
  end
  
  def do_hanoi(n, to, from, using)
    if n > 0
      do_hanoi(n-1, using, from, to)
      @moves += 1
      do_hanoi(n-1, to, using, from)
    end
  end
  
  def hanoi(n)
    do_hanoi(n, 3, 1, 2)
    puts "#@moves moves total"
  end
end

Hanoi.new.hanoi(20)
